tests/test_dataset_noResample.py

```diff
@@ -0,0 +1,485 @@
+import pytest
+import random
+import os
+import webdataset as wds
+import glob
+from open_lm.model import _MODEL_CONFIGS
+from open_lm.main import random_seed
+from open_lm.data import get_wds_dataset
+from open_lm.file_utils import (
+    get_string_for_epoch,
+    get_metadata_file,
+    get_shards_for_chunk,
+)
+from open_lm.params import parse_args
+from pathlib import Path
+from tests.utils import download_dl_test_data, make_fake_tarfiles
+import json
+from collections import Counter, defaultdict
+
+import numpy as np
+
+SEQ_LEN = 2048
+INPUT_PATHS = [
+    "tests/assets/source_id_00/manifest.jsonl",
+    "tests/assets/source_id_01/manifest.jsonl",
+]
+SINGLE_SOURCE = [INPUT_PATHS[0]]
+
+TOTAL_SEQ_COUNT = 666 * 2
+SINGLE_SEQ_COUNT = 666
+
+""" TEST STRATEGY:
+Tests to just handle the edge cases in sampling WITHOUT replacement.
+
+We are just going to cover the path where we have:
+- Exactly one source 
+- Never request more data than we have available
+
+And generally we want to assert that:
+- We never see any repeated data 
+- We see exactly the right number of samples as desired (some math here)
+
+"""
+
+
+
+# ===============================================
+# =                Helpers                      =
+# ===============================================
+
+
+
+def get_dataset_size(paths, count_tokens=False):
+    """ Computes dataset size (given list of paths to the manifest)
+        Returns number of sequences by default, but can multiply by 2048 if you ask it to =)
+    """
+    total_seq_count = 0
+    for filename in paths:
+        for line in open(filename, 'r').readlines():
+            total_seq_count += json.loads(line)['num_chunks']
+    return total_seq_count * 2048 if count_tokens else total_seq_count
+
+
+
+def retrieve_dataset_once(total_seqs, paths, epoch, next_shard, weights, seed, disable_buffer, batch_size, num_workers, 
+                          min_shards_needed=1):
+    """ Returns the output of get_wds_dataset -- not dataloader or nothing fancy
+    ONLY WORKS FOR A SINGLE SOURCE
+    """
+
+    assert len(paths) == 1
+    args = parse_args("")
+    random_seed(seed)
+
+
+    train_data_string_per_source, num_seqs_per_source, _ = get_string_for_epoch(
+        total_seqs, [next_shard], paths, weights, min_shards_needed, world_size=1)
+
+    args.train_num_samples = total_seqs
+    args.train_data = train_data_string_per_source
+    args.workers = num_workers
+    args.batch_size = batch_size
+    args.seed = seed
+    args.dataset_resampled = False
+    args.disable_buffer = disable_buffer
+    args.vocab_size = _MODEL_CONFIGS[args.model]["vocab_size"]
+    args.seq_len = _MODEL_CONFIGS[args.model]["seq_len"]
+    args.world_size = 1 
+    data = get_wds_dataset(args, is_train=True, epoch=epoch, force_num_samples=[total_seqs])
+    return data
+
+
+def retrieve_dataset_once_2(total_seqs, paths, epoch, next_shard, weights, seed, disable_buffer, batch_size, num_workers, 
```

Comment:
maybe rename to a more informative test name?

Reply:
Changed it to `retrieve_dataset_once_multi_source`.